### PR TITLE
F/wpc 773-polls-editor

### PR DESCRIPTION
**Pull request by michael-scheurer**

### CODE REVIEW

File: packages/api/src/graphql/poll/poll.private-mutation.ts [hunk 1]
```diff
@@ -178,8 +181,8 @@ export const deletePollAnswer = async (
     throw new NotFound('PollAnswer', answerId)
   }
 
-  if (answerWithPoll.poll.answers.length <= 2) {
-    throw new GraphQLError('A poll requires atleast 2 answers')
+  if (answerWithPoll.poll.answers.length <= 1) {
+    throw new GraphQLError('A poll requires at least one answer.')
```
Comment (Itrulia) [inline packages/api/src/graphql/poll/poll.private-mutation.ts:185]: Any explanation for this change? :D

File: packages/editor/src/client/atoms/poll/deletePollModal.tsx [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+  setPoll(poll: FullPoll | undefined): void
+  afterDelete(): Promise<ApolloQueryResult<PollsQuery>>
+}
+
+export function DeletePollModal({poll, setPoll, afterDelete}: deletePollProps) {
+  const {t} = useTranslation()
+  const [deletePollMutation] = useDeletePollMutation()
+  const [open, setOpen] = useState<boolean>(false)
+
+  useEffect(() => {
```
Comment (Itrulia) [inline packages/editor/src/client/atoms/poll/deletePollModal.tsx:19]: We probably should not have an useEffect here and just have open default to true.
We are not gaining much here but a double rerender.

Let's rather not render anything if poll does not exist.

File: packages/editor/src/client/atoms/poll/pollAnswers.tsx [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+  }
+
+  /**
+   * FUNCTIONS
+   */
+  async function createAnswer() {
+    if (!poll) {
+      return
+    }
+    if (newAnswer === '') {
```
Comment (Itrulia) [inline packages/editor/src/client/atoms/poll/pollAnswers.tsx:53]:
```suggestion
    if (!newAnswer) {
```

As we want to check if newAnswer is empty, could lead to a bug and opens up the question when reading the code why only empty string.

File: packages/editor/src/client/atoms/poll/pollAnswers.tsx [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+    const updatedPoll = {
+      ...poll,
+      answers: updatedAnswers
+    }
+    await onPollUpdated(updatedPoll)
+  }
+
+  return (
+    <>
+      {/* iterate poll answers */}
```
Comment (Itrulia) [inline packages/editor/src/client/atoms/poll/pollAnswers.tsx:108]: Don't think we need this comment

File: packages/editor/src/client/atoms/poll/pollAnswers.tsx [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+      onError: onErrorToast
+    })
+    await onAnswerDeleted()
+  }
+
+  async function updateAnswer(updatedAnswer: PollAnswerWithVoteCount) {
+    if (!poll) {
+      return
+    }
+    const updatedAnswers = [...(poll.answers || [])]
```
Comment (Itrulia) [inline packages/editor/src/client/atoms/poll/pollAnswers.tsx:92]:
```suggestion
    const updatedAnswers = poll.answers ? [...poll.answers] : []
```

A bit more readable for me

File: packages/editor/src/client/atoms/poll/pollExternalVotes.tsx [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+  const {t} = useTranslation()
+  const [newSource, setNewSource] = useState<string | undefined>(undefined)
+  const [sourceToDelete, setSourceToDelete] = useState<PollExternalVoteSource | undefined>(
+    undefined
+  )
+  const [openModal, setOpenModal] = useState<boolean>(false)
+  const [createSourceMutation, {loading}] = useCreatePollExternalVoteSourceMutation({
+    fetchPolicy: 'no-cache'
+  })
+  const [deletePollMutation] = useDeletePollExternalVoteSourceMutation({
```
Comment (Itrulia) [inline packages/editor/src/client/atoms/poll/pollExternalVotes.tsx:42]: I think this should be renamed as the name indicates the poll is being deleted.

File: packages/editor/src/client/atoms/poll/pollExternalVotes.tsx [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+  onExternalSourceChange(externalVoteSources: PollExternalVoteSource[]): void
+  savePoll(): Promise<void>
+}
+
+export function PollExternalVotes({
+  poll,
+  onExternalSourceCreated,
+  onExternalSourceDeleted,
+  onExternalSourceChange,
+  savePoll
```
Comment (Itrulia) [inline packages/editor/src/client/atoms/poll/pollExternalVotes.tsx:31]: Is this to fix that when the poll is not saved that we don't override it when we create a new external vote?
I think this is a bit hacky and is an unexpected user flow.

We probably should use a lazy query on the parent so the poll doesn't get updated all the time and thus overriding the data.

File: packages/editor/src/client/atoms/poll/pollExternalVotes.tsx [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+    source.voteAmounts[voteIndex].amount =
+      typeof newAmount === 'string' ? parseInt(newAmount) : newAmount
+
+    onExternalSourceChange(newVoteSources)
+  }
+
+  /**
+   * UI HELPER FUNCTIONS
+   */
+  function tableView() {
```
Comment (Itrulia) [inline packages/editor/src/client/atoms/poll/pollExternalVotes.tsx:147]: Could we make this pure (or even better a component) and move it out of the component?

File: packages/editor/src/client/atoms/poll/pollExternalVotes.tsx [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+                />
+              )}
+            </Table.Cell>
+          </Table.Column>
+        </Table>
+      </>
+    )
+  }
+
+  function iterateAnswerColumns() {
```
Comment (Itrulia) [inline packages/editor/src/client/atoms/poll/pollExternalVotes.tsx:181]: Same as above

File: packages/editor/src/client/atoms/poll/pollExternalVotes.tsx [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+                changeSource(answer, externalVoteSource, newValue)
+              }}
+            />
+          )}
+        </Table.Cell>
+      </Table.Column>
+    ))
+  }
+
+  function addSourceView() {
```
Comment (Itrulia) [inline packages/editor/src/client/atoms/poll/pollExternalVotes.tsx:203]: Same as above

File: packages/editor/src/client/atoms/poll/pollExternalVotes.tsx [hunk 11]
```diff
@@ -0,0 +1,265 @@
...
+            <IconButton icon={<PlusIcon />} appearance="primary" onClick={createPoll}>
+              {t('pollExternalVotes.addSourceBtn')}
+            </IconButton>
+          </Col>
+        </Row>
+      </>
+    )
+  }
+
+  function deleteModalView() {
```
Comment (Itrulia) [inline packages/editor/src/client/atoms/poll/pollExternalVotes.tsx:225]: Same as above

File: packages/editor/src/client/atoms/poll/pollExternalVotes.tsx [hunk 12]
```diff
@@ -0,0 +1,265 @@
...
+            </Button>
+          </Modal.Footer>
+        </Modal>
+      </>
+    )
+  }
+
+  return (
+    <>
+      {/* table */}
```
Comment (Itrulia) [inline packages/editor/src/client/atoms/poll/pollExternalVotes.tsx:257]: Don't think we need those comments

File: packages/editor/src/client/base.tsx [hunk 13]
```diff
@@ -158,7 +170,7 @@ export function Base({children}: BaseProps) {
                   {t('navbar.authors')}
                 </Nav.Item>
 
-                <Nav.Menu eventKey={'1'} title={t('navbar.usersAndMembers')} icon={<PeoplesIcon />}>
+                <Nav.Menu eventKey={'2'} title={t('navbar.usersAndMembers')} icon={<PeoplesIcon />}>
```
Comment (Itrulia) [inline packages/editor/src/client/base.tsx:173]: Out of interest, why the change? I also have no clue what eventKey does, as nothing changes when I press 1 or 2 :D

File: packages/editor/src/client/routes/polls/pollEditView.tsx [hunk 14]
```diff
@@ -0,0 +1,256 @@
...
+  const params = useParams()
+  const navigate = useNavigate()
+  const [poll, setPoll] = useState<FullPoll | undefined>(undefined)
+  const [close, setClose] = useState<boolean>(false)
+  const closePath = '/polls'
+
+  /**
+   * Handling toasts
+   */
+  const onErrorToast = (error: ApolloError) => {
```
Comment (Itrulia) [inline packages/editor/src/client/routes/polls/pollEditView.tsx:30]: Could we move these functions out of the component?

File: packages/editor/src/client/routes/polls/pollEditView.tsx [hunk 15]
```diff
@@ -0,0 +1,256 @@
...
+    onError: onErrorToast,
+    onCompleted: onCompletedToast
+  })
+  const {t} = useTranslation()
+  const loading = createLoading || updateLoading
+
+  /**
+   * Update poll object after fetching from api
+   */
+  useEffect(() => {
```
Comment (Itrulia) [inline packages/editor/src/client/routes/polls/pollEditView.tsx:65]: using `useState` instead of `useMemo` causes a double render here, can we switch to useMemo instead?

File: packages/editor/src/client/routes/polls/pollEditView.tsx [hunk 16]
```diff
@@ -0,0 +1,256 @@
...
+   */
+  async function saveOrUpdate(): Promise<void> {
+    if (!poll) {
+      return
+    }
+    const opensAt = poll.opensAt ? new Date(poll.opensAt).toISOString() : null
+    const closedAt = poll.closedAt ? new Date(poll.closedAt).toISOString() : null
+    const externalSources = poll.externalVoteSources?.map((voteSource: PollExternalVoteSource) => ({
+      ...voteSource,
+      __typename: undefined,
```
Comment (Itrulia) [inline packages/editor/src/client/routes/polls/pollEditView.tsx:100]: could also do `{__typename, ...voteSource}: PollExternalVoteSource` above

File: packages/editor/src/client/routes/polls/pollEditView.tsx [hunk 17]
```diff
@@ -0,0 +1,256 @@
...
+                    }}
+                  />
+                </Panel>
+              </Col>
+              {/* poll external votes */}
+              <Col xs={24}>
+                <Panel header={t('pollEditView.pollExternalVotesPanelHeader')} bordered>
+                  <PollExternalVotes
+                    poll={poll}
+                    onExternalSourceCreated={async () => {
```
Comment (Itrulia) [inline packages/editor/src/client/routes/polls/pollEditView.tsx:234]: don't think we need async/await here

File: packages/editor/src/client/routes/polls/pollList.tsx [hunk 18]
```diff
@@ -0,0 +1,200 @@
...
+    refetch({
+      take: limit,
+      skip: (page - 1) * limit
+    })
+  }, [page, limit])
+
+  /**
+   * UI HELPERS
+   */
+  function pollStateView(poll: FullPoll) {
```
Comment (Itrulia) [inline packages/editor/src/client/routes/polls/pollList.tsx:57]: Can we extract these functions as components?

File: packages/editor/src/client/routes/polls/pollList.tsx [hunk 19]
```diff
@@ -0,0 +1,200 @@
...
+    const now = new Date()
+    const closedAt = poll.closedAt ? new Date(poll.closedAt) : undefined
+
+    // poll has been closed
+    if (closedAt && now.getTime() >= closedAt.getTime()) {
+      return (
+        <>
```
Comment (Itrulia) [inline packages/editor/src/client/routes/polls/pollList.tsx:64]: Fragments in this function are not needed

File: packages/editor/src/client/routes/polls/pollEditView.tsx [hunk 20]
```diff
@@ -0,0 +1,256 @@
...
+                  <Form.Group controlId="question">
+                    <Form.Control
+                      name="question"
+                      placeholder={t('pollEditView.toBeOrNotToBe')}
+                      value={poll?.question || ''}
+                      onChange={(value: string) => {
+                        if (!poll) {
+                          return
+                        }
+                        setPoll({...poll, question: value})
```
Comment (Itrulia) [inline packages/editor/src/client/routes/polls/pollEditView.tsx:165]: setState's inside of event handlers in React 17 are batched (in React 18 all setStates), thus this can lead to a bug as `poll` could be stale.

```suggestion
                        setPoll(p => ({...p, question: value}))
```

Would fix this issue.

**Review by Itrulia — CHANGES_REQUESTED** [review]
It's a lot of comments but don't worry overall this has been a very good PR. I have been commenting a lot on coding style that we currently have in our code base, so they are not really blocking by me but I would love to have them fixed going forward with new code, as over time this would improve the code base.

A few things I would like to have changed:

1. There's quite a few comments that we don't really need as the line below them describes already what is going on.

2. There are inline functions inside components that render other components. Those should be extracted as seperate components. Having them as functions inline makes them not reusable (and thus not *unit* testable) and depending on some compiler shenanigans (they are sometimes inlined or not) a memory issue as they get recreated every time.

3. There's an unexpected save for the user when creating a new external vote source. This is simply fixing a symptom and not the problem that the query requeries all the time and overrides the form state. Either we should have the form state fix and not update on the users behalf, or we should make the query a lazy query. This fixes the root cause of this issue.

File: packages/api/src/graphql/poll/poll.private-mutation.ts (hunk 1 above)
Comment (michael-scheurer) [reply packages/api/src/graphql/poll/poll.private-mutation.ts:185]: Yes. Required by @LaurentAe.

File: packages/editor/src/client/base.tsx (hunk 13 above)
Comment (michael-scheurer) [reply packages/editor/src/client/base.tsx:173]: Actually, I don't know, why that changed.

File: packages/editor/src/client/atoms/poll/pollExternalVotes.tsx (hunk 8 above)
Comment (michael-scheurer) [reply packages/editor/src/client/atoms/poll/pollExternalVotes.tsx:147]: These functions are closely connected to it's component. It would not make any sense to export them as a new component. Rather it would generate boilerplate and complicate things up.

I just split this code into three parts, so that the code structure is more clear and clean and one line of code within a function has the same level of complexity.
image

File: packages/editor/src/client/routes/polls/pollEditView.tsx (hunk 16 above)
Comment (michael-scheurer) [reply packages/editor/src/client/routes/polls/pollEditView.tsx:100]: would lead to "TS18004: No value exists in scope for the shorthand property '__typename'. Either declare one or provide an initializer."

File: packages/editor/src/client/routes/polls/pollList.tsx (hunk 18 above)
Comment (michael-scheurer) [reply packages/editor/src/client/routes/polls/pollList.tsx:57]: yes. that's a good one to do so ;)

File: packages/editor/src/client/routes/polls/pollList.tsx (hunk 19 above)
Comment (michael-scheurer) [reply packages/editor/src/client/routes/polls/pollList.tsx]: Yep. And added a tooltip too.

File: packages/editor/src/client/routes/polls/pollEditView.tsx (hunk 16 above)
Comment (Itrulia) [reply packages/editor/src/client/routes/polls/pollEditView.tsx:100]: Oh is __typename not in PollExternalVoteSource?

File: packages/editor/src/client/atoms/poll/pollAnswers.tsx [hunk 21]
```diff
@@ -66,7 +58,8 @@ export function PollAnswers({
     })
     const savedAnswer = answer?.data?.createPollAnswer
     if (savedAnswer) {
-      await onAnswerAdded(savedAnswer)
+      poll.answers?.push(savedAnswer as PollAnswerWithVoteCount)
```
Comment (Itrulia) [inline packages/editor/src/client/atoms/poll/pollAnswers.tsx:61]: This is mutating a parent object and shouldn't be done.

File: packages/editor/src/client/atoms/poll/pollExternalVotes.tsx (hunk 7 above)
Comment (michael-scheurer) [reply packages/editor/src/client/atoms/poll/pollExternalVotes.tsx:31]: changed.

**Review by Itrulia — APPROVED** [review]
Looks great to me!
